File: telemetry/aitf/sdk/go/ocsf/crosswalk_test.go
package ocsf

import (
	"testing"

	"github.com/girdav01/AITF/sdk/go/semconv"
)

func TestNormalizeAgentTypeID(t *testing.T) {
	cases := []struct {
		framework string
		want      int
	}{
		{"langchain", AgentTypeIDLangChain},
		{"langgraph", AgentTypeIDLangChain},
		{"crewai", AgentTypeIDCrewAI},
		{"autogen", AgentTypeIDAutoGen},
		{"native", AgentTypeIDNative},
		{"semantic_kernel", AgentTypeIDOther},
		{"custom", AgentTypeIDOther},
		{"", AgentTypeIDUnknown},
	}
	for _, c := range cases {
		if got := NormalizeAgentTypeID(c.framework); got != c.want {
			t.Errorf("NormalizeAgentTypeID(%q) = %d, want %d", c.framework, got, c.want)
		}
	}
}

func TestBuildAIAgentCrewAI(t *testing.T) {
	attrs := map[string]interface{}{
		string(semconv.AgentIDKey):        "agent-123",
		string(semconv.AgentNameKey):      "researcher",
		string(semconv.AgentFrameworkKey): "crewai",
		string(semconv.GenAIRequestModelKey): "gpt-4o",
	}
	agent := BuildAIAgent(attrs)
	if agent == nil {
		t.Fatal("BuildAIAgent returned nil")
	}
	if agent.UID != "agent-123" {
		t.Errorf("UID = %q, want agent-123", agent.UID)
	}
	if agent.TypeID != AgentTypeIDCrewAI {
		t.Errorf("TypeID = %d, want %d", agent.TypeID, AgentTypeIDCrewAI)
	}
	if agent.Type != "CrewAI" {
		t.Errorf("Type = %q, want CrewAI", agent.Type)
	}
	if agent.AIModel != "gpt-4o" {
		t.Errorf("AIModel = %q, want gpt-4o", agent.AIModel)
	}
}

func TestBuildAIAgentNilWithoutIdentity(t *testing.T) {
	if agent := BuildAIAgent(map[string]interface{}{}); agent != nil {
		t.Errorf("BuildAIAgent on empty attrs = %+v, want nil", agent)
	}
}

func TestBuildDelegation(t *testing.T) {
	attrs := map[string]interface{}{
		string(semconv.IdentityDelegDelegateeIDKey): "delegatee-id",
		string(semconv.IdentityDelegDelegatorIDKey): "delegator-id",
		string(semconv.IdentityDelegTypeKey):        "on_behalf_of",
		string(semconv.IdentityProviderKey):         "issuer-x",
		string(semconv.IdentityDelegScopeDelegatedKey): []string{"read", "write"},
		string(semconv.IdentityDelegTTLSecondsKey):  int64(3600),
	}
	d := BuildDelegation(attrs)
	if d == nil {
		t.Fatal("BuildDelegation returned nil")
	}
	if d.UID != "delegatee-id" {
		t.Errorf("UID = %q, want delegatee-id", d.UID)
	}
	if d.ParentUID != "delegator-id" {
		t.Errorf("ParentUID = %q, want delegator-id", d.ParentUID)
	}
	if d.IssuerUID != "issuer-x" {
		t.Errorf("IssuerUID = %q, want issuer-x", d.IssuerUID)
	}
	if d.Type != "on_behalf_of" {
		t.Errorf("Type = %q, want on_behalf_of", d.Type)
	}
	if len(d.Scope) != 2 || d.Scope[0] != "read" || d.Scope[1] != "write" {
		t.Errorf("Scope = %v, want [read write]", d.Scope)
	}
	if d.TTLSeconds == nil || *d.TTLSeconds != 3600 {
		t.Errorf("TTLSeconds = %v, want 3600", d.TTLSeconds)
	}
}

func TestBuildDelegationNilWithoutContext(t *testing.T) {
	if d := BuildDelegation(map[string]interface{}{}); d != nil {
		t.Errorf("BuildDelegation on empty attrs = %+v, want nil", d)
	}
}

func TestBuildDelegationLineage(t *testing.T) {
	attrs := map[string]interface{}{
		string(semconv.IdentityDelegChainKey): []string{"root", "mid", "leaf"},
	}
	lineage := BuildDelegationLineage(attrs)
	if lineage == nil {
		t.Fatal("BuildDelegationLineage returned nil")
	}
	if len(lineage.Nodes) != 3 {
		t.Fatalf("len(Nodes) = %d, want 3", len(lineage.Nodes))
	}
	if lineage.Nodes[0].ParentUID != "" || lineage.Nodes[0].Depth != 0 {
		t.Errorf("node[0] = %+v, want empty parent and depth 0", lineage.Nodes[0])
	}
	if lineage.Nodes[2].ParentUID != "mid" || lineage.Nodes[2].Depth != 2 {
		t.Errorf("node[2] = %+v, want parent mid and depth 2", lineage.Nodes[2])
	}
	if lineage.Nodes[1].AgentUID != "mid" {
		t.Errorf("node[1].AgentUID = %q, want mid", lineage.Nodes[1].AgentUID)
	}
}

func TestCrosswalkTables(t *testing.T) {
	if got := OCSFAgentActivityCrosswalk[3]; got != "Update" { // activity 3 -> Step Execute
		t.Errorf("OCSFAgentActivityCrosswalk[3] = %q, want Update", got)
	}
	if got := OCSFAgentActivityCrosswalk[1]; got != "Spawn" {
		t.Errorf("OCSFAgentActivityCrosswalk[1] = %q, want Spawn", got)
	}
	if got := OCSFDelegationActivityCrosswalk["grant"]; got != "Create" {
		t.Errorf("OCSFDelegationActivityCrosswalk[grant] = %q, want Create", got)
	}

	entry, ok := OCSFClassCrosswalk["agent_activity"]
	if !ok {
		t.Fatal("OCSFClassCrosswalk missing agent_activity")
	}
	// AITF <= 0.4 routed this to agent_activity (9001) in a proposed "ai"
	// category. OCSF categories.json stops at uid 8, so it now reuses API
	// Activity with the ai_operation profile.
	if entry.OCSFClass != "api_activity" {
		t.Errorf("agent_activity OCSFClass = %q, want api_activity", entry.OCSFClass)
	}
	if entry.OCSFClassUID != ClassUIDAPIActivity || ClassUIDAPIActivity != 6003 {
		t.Errorf("agent_activity class_uid = %d, want 6003", entry.OCSFClassUID)
	}
	if entry.OCSFCategoryUID != OCSFCategoryUIDApplication {
		t.Errorf("agent_activity category = %d, want 6", entry.OCSFCategoryUID)
	}
	if d := OCSFClassCrosswalk["delegation_activity"]; d.OCSFClassUID != ClassUIDAuthorizeSession || d.OCSFClass != "authorize_session" {
		t.Errorf("delegation_activity = %d/%q, want 3003/authorize_session", d.OCSFClassUID, d.OCSFClass)
	}
	// No row may target the unratified ai category (uid 9).
	for name, e := range OCSFClassCrosswalk {
		if e.OCSFCategoryUID >= 9 || e.OCSFClassUID >= 9000 {
			t.Errorf("%s targets unratified category/class %d/%d", name, e.OCSFCategoryUID, e.OCSFClassUID)
		}
	}
	// Legacy UIDs stay decodable for telemetry recorded by AITF <= 0.4.
	if LegacyAIClassUIDs[9001] != ClassUIDAPIActivity || LegacyAIClassUIDs[9002] != ClassUIDAuthorizeSession || LegacyAIClassUIDs[9003] != ClassUIDAPIActivity {
		t.Errorf("LegacyAIClassUIDs = %v, want 9001/9003->6003, 9002->3003", LegacyAIClassUIDs)
	}

	// Inference and tool execution intentionally share API Activity (6003).
	inf := OCSFClassCrosswalk["model_inference"]
	tool := OCSFClassCrosswalk["tool_execution"]
	if inf.OCSFClassUID != 6003 || tool.OCSFClassUID != 6003 {
		t.Errorf("model_inference=%d, tool_execution=%d, both want 6003", inf.OCSFClassUID, tool.OCSFClassUID)
	}
	if inf.OCSFCategoryUID != 6 || tool.OCSFCategoryUID != 6 {
		t.Errorf("model_inference cat=%d, tool_execution cat=%d, both want 6", inf.OCSFCategoryUID, tool.OCSFCategoryUID)
	}
}

File: telemetry/aitf/sdk/typescript/src/ocsf/crosswalk.test.ts
/**
 * Tests for the AITF <-> OCSF agentic crosswalk (OCSF PR #1641 / issue #1640).
 */

import { ReadableSpan } from "@opentelemetry/sdk-trace-base";
import {
  buildAiAgent,
  buildDelegation,
  buildDelegationLineage,
  OCSF_AGENT_ACTIVITY_CROSSWALK,
  OCSF_CLASS_CROSSWALK,
} from "./crosswalk";
import { AgentTypeID, OCSF_AI_CATEGORY_UID, normalizeAgentTypeId } from "./schema";
import { OCSFMapper } from "./mapper";
import {
  AgentAttributes,
  IdentityAttributes,
} from "../semantic-conventions/attributes";

/** Build a minimal fake ReadableSpan for the mapper. */
function fakeSpan(
  name: string,
  attributes: Record<string, unknown>
): ReadableSpan {
  return {
    name,
    attributes,
    startTime: [1_700_000_000, 0],
  } as unknown as ReadableSpan;
}

describe("normalizeAgentTypeId", () => {
  it("maps langchain to LangChain (2)", () => {
    expect(normalizeAgentTypeId("langchain")).toBe(AgentTypeID.LANGCHAIN);
  });

  it("maps langgraph to LangChain (2)", () => {
    expect(normalizeAgentTypeId("langgraph")).toBe(AgentTypeID.LANGCHAIN);
  });

  it("maps crewai to CrewAI (4)", () => {
    expect(normalizeAgentTypeId("crewai")).toBe(AgentTypeID.CREWAI);
  });

  it("maps an unknown non-empty framework to Other (99)", () => {
    expect(normalizeAgentTypeId("semantic_kernel")).toBe(AgentTypeID.OTHER);
  });

  it("maps undefined/empty to Unknown (0)", () => {
    expect(normalizeAgentTypeId(undefined)).toBe(AgentTypeID.UNKNOWN);
    expect(normalizeAgentTypeId("")).toBe(AgentTypeID.UNKNOWN);
  });
});

describe("buildAiAgent", () => {
  it("returns undefined when no agent identity is present", () => {
    expect(buildAiAgent({})).toBeUndefined();
  });

  it("builds type_id=4 / type=CrewAI for a crewai agent", () => {
    const agent = buildAiAgent({
      [AgentAttributes.ID]: "agent-1",
      [AgentAttributes.NAME]: "Researcher",
      [AgentAttributes.FRAMEWORK]: "crewai",
    });
    expect(agent).toBeDefined();
    expect(agent!.uid).toBe("agent-1");
    expect(agent!.name).toBe("Researcher");
    expect(agent!.type_id).toBe(AgentTypeID.CREWAI);
    expect(agent!.type).toBe("CrewAI");
  });
});

describe("buildDelegation", () => {
  it("returns undefined when no delegation context is present", () => {
    expect(buildDelegation({})).toBeUndefined();
  });

  it("uses delegatee_id as uid and delegator_id as parent_uid", () => {
    const delegation = buildDelegation({
      [IdentityAttributes.DELEGATION_DELEGATEE_ID]: "delegatee-9",
      [IdentityAttributes.DELEGATION_DELEGATOR_ID]: "delegator-3",
      [IdentityAttributes.DELEGATION_SCOPE_DELEGATED]: ["read", "write"],
    });
    expect(delegation).toBeDefined();
    expect(delegation!.uid).toBe("delegatee-9");
    expect(delegation!.parent_uid).toBe("delegator-3");
    expect(delegation!.scope).toEqual(["read", "write"]);
  });
});

describe("buildDelegationLineage", () => {
  it("materializes the delegation chain into a directed graph", () => {
    const lineage = buildDelegationLineage({
      [IdentityAttributes.DELEGATION_CHAIN]: ["a", "b", "c"],
    });
    expect(lineage).toBeDefined();
    expect(lineage!.nodes).toHaveLength(3);
    expect(lineage!.nodes[0]).toMatchObject({
      uid: "a",
      parent_uid: undefined,
      depth: 0,
    });
    expect(lineage!.nodes[2]).toMatchObject({
      uid: "c",
      parent_uid: "b",
      depth: 2,
    });
  });
});

describe("OCSFMapper ai_operation enrichment", () => {
  it("attaches event.ai_agent to a mapped agent span", () => {
    const mapper = new OCSFMapper();
    const span = fakeSpan("agent.step.execute", {
      [AgentAttributes.NAME]: "Planner",
      [AgentAttributes.ID]: "agent-42",
      [AgentAttributes.FRAMEWORK]: "langchain",
      [AgentAttributes.SESSION_ID]: "conv-1",
    });
    const event = mapper.mapSpan(span);
    expect(event).not.toBeNull();
    expect(event!.ai_agent).toBeDefined();
    expect(event!.ai_agent!.uid).toBe("agent-42");
    expect(event!.ai_agent!.type_id).toBe(AgentTypeID.LANGCHAIN);
    expect(event!.ai_agent!.type).toBe("LangChain");
  });
});

describe("crosswalk tables", () => {
  it("reuses existing OCSF classes for data-plane events", () => {
    expect(OCSF_CLASS_CROSSWALK["model_inference"]).toEqual({
      ocsf_category_uid: 6,
      ocsf_class_uid: 6003,
      ocsf_class: "api_activity",
    });
    // Inference and tool execution both reuse API Activity (6003).
    expect(OCSF_CLASS_CROSSWALK["tool_execution"].ocsf_class_uid).toBe(6003);
    expect(OCSF_CLASS_CROSSWALK["data_retrieval"].ocsf_class_uid).toBe(6005);
    expect(OCSF_CLASS_CROSSWALK["security_finding"].ocsf_class_uid).toBe(2004);
    expect(OCSF_CLASS_CROSSWALK["supply_chain"].ocsf_class_uid).toBe(2002);
    expect(OCSF_CLASS_CROSSWALK["governance"].ocsf_class_uid).toBe(2003);
    expect(OCSF_CLASS_CROSSWALK["identity"].ocsf_class_uid).toBe(3002);
    expect(OCSF_CLASS_CROSSWALK["model_ops"].ocsf_class_uid).toBe(6002);
    expect(OCSF_CLASS_CROSSWALK["asset_inventory"].ocsf_class_uid).toBe(5001);
  });

  it("routes control-plane classes onto released OCSF classes", () => {
    // The proposed `ai` category (uid 9) was never ratified: OCSF
    // categories.json stops at uid 8 (unmanned_systems).
    expect(OCSF_AI_CATEGORY_UID).toBeNull();
    expect(OCSF_CLASS_CROSSWALK["agent_activity"]).toEqual({
      ocsf_category_uid: 6,
      ocsf_class_uid: 6003,
      ocsf_class: "api_activity",
    });
    expect(OCSF_CLASS_CROSSWALK["delegation_activity"]).toEqual({
      ocsf_category_uid: 3,
      ocsf_class_uid: 3003,
      ocsf_class: "authorize_session",
    });
    expect(OCSF_CLASS_CROSSWALK["agent_communication"]).toEqual({
      ocsf_category_uid: 6,
      ocsf_class_uid: 6003,
      ocsf_class: "api_activity",
    });
    // No row may target the unratified ai category.
    for (const target of Object.values(OCSF_CLASS_CROSSWALK)) {
      expect(target.ocsf_category_uid).toBeLessThan(9);
      expect(target.ocsf_class_uid).toBeLessThan(9000);
    }
  });

  it("collapses to 9 distinct reused class_uids", () => {
    const classUids = new Set(
      Object.values(OCSF_CLASS_CROSSWALK).map((t) => t.ocsf_class_uid)
    );
    expect(classUids).toEqual(
      new Set([6003, 6005, 6002, 2004, 2002, 2003, 3002, 3003, 5001])
    );
  });

  it("maps agent activity_id 1 (Session Start) to Spawn", () => {
    expect(OCSF_AGENT_ACTIVITY_CROSSWALK[1]).toBe("Spawn");
    expect(OCSF_AGENT_ACTIVITY_CROSSWALK[99]).toBe("Unknown");
  });
});
